kimchi/src/proof.rs

```diff
@@ -2,126 +2,225 @@
 
 use crate::circuits::wires::{COLUMNS, PERMUTS};
 use ark_ec::AffineCurve;
-use ark_ff::{FftField, Zero};
+use ark_ff::{FftField, Field, Zero};
 use ark_poly::univariate::DensePolynomial;
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
 use array_init::array_init;
 use commitment_dlog::{commitment::PolyComm, evaluation_proof::OpeningProof};
 use o1_utils::{types::fields::*, ExtendedDensePolynomial};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use serde_with::serde_as;
 
 //~ spec:startcode
-#[derive(Clone)]
-pub struct LookupEvaluations<Field> {
+#[serde_as]
+#[derive(Clone, Deserialize, Serialize)]
+pub struct LookupEvaluations<F: CanonicalSerialize + CanonicalDeserialize> {
     /// sorted lookup table polynomial
-    pub sorted: Vec<Field>,
+    #[serde_as(as = "Vec<Vec<o1_utils::serialization::SerdeAs>>")]
+    pub sorted: Vec<Vec<F>>,
     /// lookup aggregation polynomial
-    pub aggreg: Field,
+    #[serde_as(as = "Vec<o1_utils::serialization::SerdeAs>")]
+    pub aggreg: Vec<F>,
     // TODO: May be possible to optimize this away?
     /// lookup table polynomial
-    pub table: Field,
-
+    #[serde_as(as = "Vec<o1_utils::serialization::SerdeAs>")]
+    pub table: Vec<F>,
     /// Optionally, a runtime table polynomial.
-    pub runtime: Option<Field>,
+    #[serde_as(as = "Option<Vec<o1_utils::serialization::SerdeAs>>")]
+    pub runtime: Option<Vec<F>>,
 }
 
 // TODO: this should really be vectors here, perhaps create another type for chunked evaluations?
-#[derive(Clone)]
-pub struct ProofEvaluations<Field> {
+#[serde_as]
+#[derive(Clone, Deserialize, Serialize)]
+pub struct ProofEvaluations<F: CanonicalSerialize + CanonicalDeserialize> {
     /// witness polynomials
-    pub w: [Field; COLUMNS],
+    #[serde_as(as = "[Vec<o1_utils::serialization::SerdeAs>; COLUMNS]")]
+    pub w: [Vec<F>; COLUMNS],
     /// permutation polynomial
-    pub z: Field,
+    #[serde_as(as = "Vec<o1_utils::serialization::SerdeAs>")]
+    pub z: Vec<F>,
     /// permutation polynomials
     /// (PERMUTS-1 evaluations because the last permutation is only used in commitment form)
-    pub s: [Field; PERMUTS - 1],
+    #[serde_as(as = "[Vec<o1_utils::serialization::SerdeAs>; PERMUTS - 1]")]
+    pub s: [Vec<F>; PERMUTS - 1],
     /// lookup-related evaluations
-    pub lookup: Option<LookupEvaluations<Field>>,
+    #[serde(bound = "LookupEvaluations<F>: Serialize")]
+    pub lookup: Option<LookupEvaluations<F>>,
     /// evaluation of the generic selector polynomial
-    pub generic_selector: Field,
+    #[serde_as(as = "Vec<o1_utils::serialization::SerdeAs>")]
+    pub generic_selector: Vec<F>,
     /// evaluation of the poseidon selector polynomial
-    pub poseidon_selector: Field,
+    #[serde_as(as = "Vec<o1_utils::serialization::SerdeAs>")]
+    pub poseidon_selector: Vec<F>,
```

Comment:
I think we should try to remove the `Vec` here and see if it works

Reply:
Just on the poseidon selector or the whole thing?

Reply:
Actually, when defining `chunked_evals` in the proof, it assumes selectors can as well be chunked.